### PR TITLE
feat: x509 certificate using garm config `v2.1.0`

### PR DESCRIPTION
**Pull request by mlajkim**
# Description

X.509 Mode Log:
```
2024-12-03 07:31:54    [INFO]:    Garm is starting with X.509 mode[true] ...                                                         
2024-12-03 07:31:54    [INFO]:    Booting X.509 Certificate reloader with .x509.cert[/var/run/athenz/service.cert.pem] .x509.key[/var/
run/athenz/test-tls.key] .x509.poll_interval[15m0s]                                                                                  
2024-12-03 07:31:54    [INFO]:    Successfully loaded X.509 certificate [/var/run/athenz/service.cert.pem] and its key [/var/run/athe
nz/test-tls.key] from local file                                                                                                     
2024-12-03 07:31:54    [WARN]:    time: invalid duration ""                                                                          
2024-12-03 07:31:54    [DEBG]:    Running garm as user [garm] with UID [100] ...                                                     
2024-12-03 07:31:54    [INFO]:    garm api server starting                                                                           
2024-12-03 07:31:54    [INFO]:    garm health check server starting 
```

Ntoken mode:
```
2024-12-03 07:38:24    [INFO]:    Garm is starting with X.509 mode[false] ...
2024-12-03 07:38:24    [WARN]:    time: invalid duration ""
2024-12-03 07:38:24    [INFO]:    Starting ntoken updater with refresh duration [30s] ...
2024-12-03 07:38:24    [DEBG]:    Running garm as user [garm] with UID [100] ...
2024-12-03 07:38:24    [INFO]:    garm health check server starting
2024-12-03 07:38:24    [INFO]:    garm api server starting
```

## TODOs
- [x] Implementation
- [x] Refactor
- [x] Handle code TODOs
- [x] Clean Logger (mimic SIA)

## Assignees
- [x] `Assignees` is set

## Type of changes
- [x] Apply one or more `labels` of the following that fits:
  - `bug`: Bug fix
  - `dependencies`: Dependency upgrades
  - `documentation`: Documentation changes
  - `enhancement`: New Feature
  - `good first issue`: First contribution
  - `logging`: Log changes
  - `refactor`: Refactoring (no functional changes, no api changes)

### Flags
```
- [ ] Breaks backward compatibility
- [ ] Requires a documentation update
- [ ] Has untestable code
```
---

## Checklist
```
- [ ] Followed the guidelines in the CONTRIBUTING document
- [ ] Added prefix `[skip ci]`/`[ci skip]`/`[no ci]`/`[skip actions]`/`[actions skip]` in the PR title if necessary
- [ ] Tested and linted the code
- [ ] Commented the code
- [ ] Made corresponding changes to the documentation
```

## Checklist for maintainer
```
- [ ] Use `Squash and merge`
- [ ] Double-confirm the merge message has prefix `[skip ci]`/`[ci skip]`/`[no ci]`/`[skip actions]`/`[actions skip]`
- [ ] Delete the branch after merge
```

### CODE REVIEW

**Comment by mlajkim** [comment]
Wanna change the log here:
```
2024-12-03 07:31:54    [INFO]:    garm api server starting                                                                           
2024-12-03 07:31:54    [INFO]:    garm health check server starting
```
=>
```
2024-12-03 07:31:54    [INFO]:    Starting Garm API server ...                                                                       
2024-12-03 07:31:54    [INFO]:    Starting Garm health check server ...
```

File: config/testdata/example_config.yaml [hunk 1]
```diff
@@ -19,6 +19,10 @@ athenz:
   url: https://www.athenz.com/zts/v1
   timeout: 5s
   root_ca: _root_ca_
+x509:
+  cert: ""
+  key: ""
+  poll_interval: ""
```
Comment (WindzCUHK) [inline config/testdata/example_config.yaml:25]:
```suggestion
athenz:
  tls/mtls/client_cert:
    cert: ""
    key: ""
    poll_interval: ""
```
I think it is better to put under Athenz struct.

File: main.go [hunk 2]
```diff
@@ -65,9 +65,21 @@ func run(cfg config.Config) []error {
 		glg.Get().DisableColor()
 	}
 
-	daemon, err := usecase.New(cfg)
-	if err != nil {
-		return []error{errors.Wrap(err, "failed to instantiate daemon")}
+	// Check if X.509 cert mode:
+	useX509Mode := cfg.X509.Cert != "" && cfg.X509.Key != ""
+	glg.Infof("Garm is starting with X.509 mode[%t] ...", useX509Mode)
```
Comment (WindzCUHK) [inline main.go:70]:
```suggestion
	if useX509Mode {
		glg.Infoln("Garm will connect Athenz server with X.509 client certificate ..."
	} else {
		glg.Infoln("Garm will connect Athenz server with n-token ...")
	}
```
more verbose

File: usecase/garmd.go [hunk 3]
```diff
@@ -66,8 +67,45 @@ func New(cfg config.Config) (GarmDaemon, error) {
 	}, nil
 }
 
+func NewX509(cfg config.Config) (GarmDaemon, error) {
+	pollInterval, err := time.ParseDuration(cfg.X509.PollInterval)
+	if err != nil {
+		return nil, errors.Wrap(err, "invalid 	config .x509.poll_interval ["+cfg.X509.PollInterval+"]")
+	}
+
+	certReloader, err := service.NewCertReloader(service.CertReloaderCfg{
```
Comment (WindzCUHK) [inline usecase/garmd.go:76]: We already have the CA config. I think it is better to handle it.
https://github.com/AthenZ/garm/blob/820f972fd56b675cde0a26d29a0b9b2f393f54d3/config/config.go#L130-L131

File: service/cert-reloader.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+)
+
+var defaultPollInterval = 1 * time.Minute
+
+// LogFn allows customized logging.
+type LogFn func(format string, args ...interface{})
+
+// CertReloader reloads the (key, cert) pair from the filesystem when
+// the cert file is updated.
+type CertReloader struct {
```
Comment (WindzCUHK) [inline service/cert-reloader.go:35]: Let's separate this as independent library in `garm v4`. Prefer not update to much from the [SIA implementation](https://github.com/AthenZ/k8s-athenz-sia/blob/main/pkg/util/cert-reloader.go).

File: config/testdata/example_config.yaml (hunk 1 above)
Comment (mlajkim) [reply config/testdata/example_config.yaml:25]:
```yaml
athenz:
  auth_header: Athenz-Principal-Auth
  url: https://www.athenz.com/zts/v1
  timeout: 5s
  cert: "" # new
  key: "" # new
  cert_key_poll_interval: "" # new
  root_ca: _root_ca_
```

File: main.go (hunk 2 above)
Comment (mlajkim) [reply main.go:70]: OK

File: usecase/garmd.go (hunk 3 above)
Comment (mlajkim) [reply usecase/garmd.go:76]: OK

File: service/cert-reloader.go (hunk 4 above)
Comment (mlajkim) [reply service/cert-reloader.go:35]: OK

File: config/testdata/example_config.yaml (hunk 1 above)
Comment (mlajkim) [reply config/testdata/example_config.yaml:25]: my bad. I will just do `poll_interval` instead. `cert_key_poll_interval` sounds weird.

File: main.go (hunk 2 above)
Comment (mlajkim) [reply main.go]: Handled in https://github.com/AthenZ/garm/pull/50/commits/2351b1c883e9f2027efbd843c2d90ff181cb5458

File: service/cert-reloader.go (hunk 4 above)
Comment (mlajkim) [reply service/cert-reloader.go:41]: Handled in https://github.com/AthenZ/garm/pull/50/commits/2f224862ef8971af729aab619335f1f0b81c407f

File: config/testdata/example_config.yaml (hunk 1 above)
Comment (mlajkim) [reply config/testdata/example_config.yaml]: Handled in https://github.com/AthenZ/garm/pull/50/commits/67e2e38fd1e4f9ff365cbca307d3657ae63f2330

**Comment by mlajkim** [comment]
The poll_interval needs to be mandatory without a default because Garm isn't aware of the expiration date.

File: usecase/garmd.go (hunk 3 above)
Comment (mlajkim) [reply usecase/garmd.go:76]: Handled. Please take a look for the entire code changes.

Comment (mlajkim) [reply usecase/garmd.go:76]: Run `loadCa()` only during the initialization

File: service/cert-reloader.go [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+		return errors.Wrap(err, fmt.Sprintf("unable to load ca cert from %s", w.caPath))
+	}
+
+	caPool, err := NewX509CertPool(w.caPath)
+	if err != nil {
+		return errors.Wrap(err, "authorization x509 certpool error")
+	}
+
+	if !caPool.AppendCertsFromPEM(caPEM) {
+		return errors.New("Certification Failed")
```
Comment (mlajkim) [inline service/cert-reloader.go:104]:
```suggestion
		return errors.New("appending ca certificate into ca pool has failed")
```

File: service/cert-reloader.go [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+func (w *CertReloader) GetCertFromCache() (*tls.Certificate, error) {
+	w.l.RLock()
+	c := w.cert
+	w.l.RUnlock()
+	return c, nil
+}
+
+// GetWebhook returns a function that is used to get X.509 Certificate stored in memory to connect to Athenz server
+// type IdentityAthenzX509 = func() (*tls.Config, error)
+func (w *CertReloader) GetWebhook() func() (*tls.Config, error) {
```
Comment (WindzCUHK) [inline service/cert-reloader.go:66]: rename?
```suggestion
func (w *CertReloader) GetTLSConfig() func() (*tls.Config, error) {
```

Comment (mlajkim) [reply service/cert-reloader.go:66]: This returns a function, not a config so no.
https://github.com/AthenZ/garm/blob/master/service/athenz_test.go#L166
Can be the following tho:
`GetAthenzWebhook()`
but yeah, not config.

Comment (WindzCUHK) [reply service/cert-reloader.go:66]: may be the following?
- `GetIdentityAthenzX509`
- `GetTLSConfigFunc`

**Comment by mlajkim** [comment]
- [ ] Use go 1.24 just incase as it seems like go 1.23 has the following error:
`fatal error: to many concurrent timer firings`

File: service/cert-reloader.go (hunk 6 above)
Comment (mlajkim) [reply service/cert-reloader.go:66]: Will do `GetTLSConfigFunc()`

**Review by WindzCUHK — APPROVED** [review]

**Comment by mlajkim** [comment]
This PR created an issue here: 
image
https://github.com/AthenZ/garm/issues/53

**Pull request merged**